nomic/aws/sagemaker.py

```diff
@@ -0,0 +1,181 @@
+# Note: This file contains methods to prepare embedding requests for Sagemaker.
+# It may make sense to move code eventually to embed.py or somewhere more generic but
+# it currently lives here to separate out dependencies.
+
+import hashlib
+import logging
+from contextlib import contextmanager
+from enum import Enum
+from pathlib import Path
+from typing import List, Optional
+
+import numpy as np
+import tritonclient.http.aio as aiohttpclient
+from tokenizers import Tokenizer
+from tritonclient.http import InferenceServerClient
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+text_embedding_model_info = {
+    "nomic-embed-text-v1.5": {
+        "dim": 768,
+        "max_length": 2048,
+        "pad_id": 0,
+        "recommended_dims": [768, 512, 384, 256, 128],
+    },
+}
+
+NULL_PLACEHOLDER = hashlib.md5(b"nomic null").hexdigest()
+EMPTY_PLACEHOLDER = hashlib.md5(b"nomic empty").hexdigest()
+
+
+class NomicTextEmbeddingModel(Enum):
+    nomic_embed_text_v1_5 = "nomic-embed-text-v1.5"
+
+    def hamming_capable(self):
+        return self == NomicTextEmbeddingModel.nomic_embed_text_v1_5
+
+    def matryoshka_capable(self):
+        return self == NomicTextEmbeddingModel.nomic_embed_text_v1_5
+
+    def dim(self):
+        return text_embedding_model_info[self.value]["dim"]
+
+    def recommended_dims(self):
+        return text_embedding_model_info[self.value].get("recommended_dims") or [
+            self.dim()
+        ]
+
+    def max_length(self):
+        return text_embedding_model_info[self.value]["max_length"]
+
+    def pad_token(self) -> int:
+        return text_embedding_model_info[self.value]["pad_id"]
+
+
+def null_empty_placeholder(text: Optional[str]) -> str:
+    """
+    If text is null or empty replace with a placeholder nonsense string.
+    """
+    if text is None:
+        return NULL_PLACEHOLDER
+    if text.strip() == "":
+        return EMPTY_PLACEHOLDER
+    return text
+
+
+def load_tokenizer(model: NomicTextEmbeddingModel) -> Tokenizer:
+    """
+    Instantiate the huggingface tokenizer.
+    :param model: the model doing the embedding
+    :return:
+    """
+    tokdir = Path(__file__).parent / "tokenizers"
+    if model == NomicTextEmbeddingModel.nomic_embed_text_v1_5:
```

Comment:
are we only supporting 1.5? what about v1?

Reply:
Just v1.5 for now. We will need to upload a new sagemaker endpoint for each version. As i understand it, we'll stick to one and get it working for marketplace/mongo before doing more